File: client/CommClient.py
#This file is distributed under the terms of the GNU General Public license.
#Copyright (C) 1999 Aloril (See the file COPYING for details).

from atlas import *
from world.physics.Vector3D import Vector3D
from transport import *
from CharacterClient import CharacterClient
from common.BaseEntity import BaseEntity
import time
from server.WorldTime import WorldTime,seconds2string

#initialize atlas world_info
world_info.coordinates_conversion_class=Vector3D
world_info.seconds2string=seconds2string
world_info.string2DateTime=WorldTime

class CommClient(BaseEntity):
    def __init__(self, argv):
        self.connection=self.connect(argv)
        playerName="player"+`randint(0,100)`
        self.player=self.create_player(playerName,password)
        self.character=self.create_character("character",CharacterClient)
    def connect(self, argv):
        if len(argv)>1:
            server=argv[1]
        else:
            server="localhost"
        if len(argv)>2:
            port=int(argv[2])
        else:
            port=6767
        try:
            return connect(server,port)
        except:
            print "Could not connect to",server,port
            raise
    def send(self, op):
        self.connection.send(op)
    def pop(self):
        return self.connection.pop()
    def create_player(self, name, password):
        player=Entity(name,password=password,type=["player"])
        createAccountOp=Operation("create",player)
        print "Create create op"
        res=self.connection.call(createAccountOp)
        print "Sent create op"
        op=res[0]
        print "Decoded reply to create op"
        if op.id!="info":
            print "account creation failed, trying to login:"
            player=Entity(name,password=password)
            loginOp=Operation("login",player)
            res=self.connection.call(loginOp)
            op=res[0]
            if op.id!="info":
                print "Error in login:\n",res
                sys.exit()
        ent=op[0]
        print "Player id:",name
        if hasattr(ent,"characters"):
            print "Characters:",ent.characters
        return ent
    def create_character(self, type, type_class):
        character=Entity(name=self.player.id,type=[type])
        createOp=Operation("create",character,from_=self.player.id)
        res=self.connection.call(createOp)
        op=res[0]
        if op.id!="info":
            raise ValueError, "creation of character of type '%s' failed" % type
        body=op[0]
        print "our character:\n",body
        obj=object_from_entity(type_class,body)
        obj.connection=self
        #CHEAT! something more elegant?
        obj.location.parent=obj.map.add(Entity(obj.location.parent,
                                               type=["thing"]))
        obj.location.coordinates=Vector3D(obj.location.coordinates)
        return obj
    def idle(self):
        time.sleep(0.1)
    def handle_net(self):
        while 1:
            input=self.connection.pop()
            if input:
                #print "Received:\n",input
                result=self.character.message(input)
                self.connection.send(result)
            else:
                break
    def run(self, loop=1):
        while 1:
            self.handle_net()
            self.idle()
            if not loop: break

File: client/CommClient.cpp
// This file may be redistributed and modified only under the terms of
// the GNU General Public License (See COPYING for details).
// Copyright (C) 2000,2001 Alistair Riddoch

#include <common/BaseEntity.h>

#include "CommClient.h"
#include "CreatorClient.h"

#include <common/debug.h>

static const bool debug_flag = false;

using Atlas::Message::Object;

CommClient::CommClient()
{
}

Object::MapType
CommClient::createPlayer(const std::string & name,
                               const std::string & password)
{
    playerName = name;
    Object::MapType player_ent;
    player_ent["username"] = name;
    player_ent["password"] = password;
    player_ent["parents"] = Object::ListType(1, "player");
    
    debug(std::cout << "Loggin " << name << " in with " << password << " as password"
               << std::endl << std::flush;);
    
    Login loginAccountOp(Login::Instantiate());
    loginAccountOp.SetArgs(Object::ListType(1,player_ent));
    send(loginAccountOp);

    if (connection.wait()) {
        Create createAccountOp(Create::Instantiate());
        createAccountOp.SetArgs(Object::ListType(1,player_ent));
        send(createAccountOp);
        if (connection.wait()) {
            std::cerr << "ERROR: Failed to log into server" << std::endl
                      << std::flush;
            return Object::MapType();
        }
    }

    Object::MapType ent = connection.getReply();

    //if (ent.find("characters") != ent.end()) {
    //}

    return ent;
}

CreatorClient * CommClient::createCharacter(const std::string & type)
{
    Object::MapType character;
    character["name"] = playerName;
    character["parents"] = Object::ListType(1,type);

    Create createOp=Create::Instantiate();
    createOp.SetFrom(playerName);
    createOp.SetArgs(Object::ListType(1,character));
    send(createOp);

    if (connection.wait()) {
        std::cerr << "ERROR: Failed to create character type: "
                  << type << std::endl << std::flush;
        return NULL;
    }
    Object::MapType body=connection.getReply();

    const std::string & id = body.find("id")->second.AsString();

    EntityDict tmp;

    CreatorClient * obj = new CreatorClient(id, type, connection);
    obj->merge(body);
    obj->getLocation(body, tmp);
    // obj = EntityFactory::instance()->newThing(type, body, tmp);
    // FIXME Do we need to create a local entity for this as is done in
    // the python version? If so, do we need to keep track of a full world
    // model here, or just in the minds (when we become an AI client
    return obj;
}

// I'm making this pure virtual, to see if that is desired.
//void CommClient::idle() {
//    time.sleep(0.1);
//}

void CommClient::handleNet()
{
    RootOperation * input;
    while ((input = connection.pop()) != NULL) {
        OpVector result = character->message(*input);
        for (OpVector::const_iterator I=result.begin(); I != result.end(); I++) {
            send(*(*I));
        }
        delete input;
    }
}
